Sync: add queue size to actions

roccotripaldi opened the pull request:
Fixes 0-1157114176685490-1161533481166414-asa

#### Changes proposed in this Pull Request:
* This PR adds the incremental queue size to requests we send to the server
* This PR also cleans up a few sync unit tests to make the test suite easier to run locally

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This enhances sync allowing us to monitor changes in queue size
* For more context see:  0-1157114176685490-1161533481166414-asa

#### Testing instructions:
* Requires a server-side patch to test
* See D39048-code for testing instructions

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* none needed

roccotripaldi committed:
Sync

Ensure we send queue size with every action from the incremental sync queue.

roccotripaldi committed:
[not verified]Sync: unit tests

- adds a unit test to test for queue size property in actions
- fixes some errors in certain sync unit tests to make them run more smoothly when run locally
- not verified, because linter has rules that should not apply to unit tests

gravityrail commented:
This is not going to work very well. `count(*)` is an expensive operation and computing it for every action added to the queue is going to slow things down a lot.

Why can't we just send the queue size when we send the actions to the server, instead of with every action?

gravityrail reviewed (dismissed):
This approach is too inefficient IMO. I need more information about why every action needs to be accompanied by the queue size.

roccotripaldi commented:
Thanks Dan! This is the exact type of feedback I was hoping for.

> Why can't we just send the queue size when we send the actions to the server, instead of with every action?

That's the desired result, but I wasn't sure where in the code base to add this. I'll do some more digging today.

roccotripaldi commented:
I've updated the description to clarify that we don't need `queue_size` attached to every action, but only to a requests sent to the sync server.

roccotripaldi committed:
Sync:

Move 'queue_size' to a $_GET var with every sync request,
instead of being attached to every sync action.

gravityrail commented:
Looks like you figured out out :) latest version is better. Thanks.

mdbitz commented on packages/sync/src/class-actions.php:
```diff
@@ -260,18 +260,20 @@ public static function set_is_importing_true() {
 	 * @return Jetpack_Error|mixed|WP_Error  The result of the sending request.
 	 */
-	public static function send_data( $data, $codec_name, $sent_timestamp, $queue_id, $checkout_duration, $preprocess_duration ) {
+	public static function send_data( $data, $codec_name, $sent_timestamp, $queue_id, $checkout_duration, $preprocess_duration, $queue_size = 0 ) {
```
I don't think we should have 0 as the default state. Having _null_ as the default will make it so that queue_size isn't recorded as 0 in the site-profiles index for older versions of jetpack. Yes we could filter by Jetpack version but null is a better representation of the data.

mdbitz reviewed (dismissed):
Let's have the default of null vs 0 :)

roccotripaldi committed:
Sync: send queue size

Defaul queue size should be null

mdbitz commented:
Leaving this in Needs Review for the moment.

I've run though testing of this change with the changes in D39048 and confirm $queue_size is available for our use cases.

Without D39048 no issues are experienced by the WP.com infrastructure. 

Looking good to merge awaiting a final regression test of D39048.

mdbitz reviewed (approved):
Reviewed and Tested with and without handlers on WP.com Queue Size flows through for Sync queue :)

mdbitz commented:
Note that no changelog should be needed for this PR as it doesn't impact end users or developers.

Internally this allows the current queue size to be sent with every sync request so we can use it as a data point for analysis and self-heal routines for sites.
